tokio-util/src/sync/mpsc.rs

```diff
@@ -0,0 +1,133 @@
+//! Tokio-02 style MPSC channel.
+use std::{
+    future::Future,
+    marker::PhantomPinned,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tokio::sync::mpsc::{error::SendError, Permit};
+
+/// Fat sender with `poll_ready` support.
+#[derive(Debug)]
+// TODO: it doesn't seem we really need 'static here...
+pub struct Sender<T: 'static> {
+    // field order: `state` can contain reference to `inner`,
+    // so it must be dropped first.
+    /// State of this sender
+    state: State<T>,
+    /// Underlying channel
+    inner: tokio::sync::mpsc::Sender<T>,
+    /// Pinned marker
+    pinned: PhantomPinned,
+}
+
+impl<T: 'static> Sender<T> {
+    /// It is OK to get mutable reference to state, because it is not
+    /// self-referencing.
+    fn pin_project_state(self: Pin<&mut Self>) -> &mut State<T> {
+        unsafe { &mut Pin::into_inner_unchecked(self).state }
+    }
+
+    /// Sender must be pinned because state can contain references to it
+    fn pin_project_inner(self: Pin<&mut Self>) -> Pin<&mut tokio::sync::mpsc::Sender<T>> {
+        unsafe { self.map_unchecked_mut(|t| &mut t.inner) }
+    }
+}
+
+enum State<T: 'static> {
+    /// We do not have permit, but we didn't start acquiring it.
+    Empty,
+    /// We have a permit to send.
+    // ALERT: this is self-reference to the Sender.
+    Ready(Permit<'static, T>),
+    /// We are in process of acquiring a permit
+    // ALERT: contained future contains self-reference to the sender.
+    Acquire(Pin<Box<dyn Future<Output = Result<Permit<'static, T>, SendError<()>>>>>),
```

Comment:
This boxing can be avoided when `type_alias_impl_trait` finally lands.

Reply:
Alternatively, if this was implemented in `tokio` rather than in `tokio-util`, this could just use the internal semaphore implementation's [`Acquire` future](https://github.com/tokio-rs/tokio/blob/90c2a510e28a05cc72f76eb5e8bd4a7f7751e333/tokio/src/sync/batch_semaphore.rs#L221). But, I don't think that's going to ever become public API, so this would need to be in `tokio` proper.